File: kr_pipeline/indicators/compute/minervini.py
# kr_pipeline/indicators/compute/minervini.py
"""미너비니 Trend Template 조건 c1-c7 계산 (c8 = rs_rating >= 70, pass = ALL 은 SQL 에서).

모든 입력은 수정종가 기준 컬럼이어야 함.
"""
import numpy as np
import pandas as pd

from kr_pipeline.common.thresholds import (
    C3_SMA200_LOOKBACK_DAYS,
    C6_W52LOW_MULT,
    C7_W52HIGH_MULT,
)


def compute_minervini_c1_to_c7(
    df: pd.DataFrame,
    sma_200_lookback: int = C3_SMA200_LOOKBACK_DAYS,
) -> pd.DataFrame:
    """
    입력 df 컬럼: adj_close, sma_50, sma_150, sma_200, w52_high, w52_low
    출력: minervini_c1 ~ minervini_c7 boolean 컬럼 (NaN 가능)

    NaN 입력 시 조건도 NaN (pandas 비교 의미).
    """
    out = pd.DataFrame(index=df.index)

    close = df["adj_close"]
    sma50 = df["sma_50"]
    sma150 = df["sma_150"]
    sma200 = df["sma_200"]
    w52h = df["w52_high"]
    w52l = df["w52_low"]

    # C1: close > sma_150 > sma_200
    out["minervini_c1"] = (close > sma150) & (sma150 > sma200)
    # C2: sma_150 > sma_200
    out["minervini_c2"] = sma150 > sma200
    # C3: sma_200(today) > sma_200(today - 22)
    sma200_lagged = sma200.shift(sma_200_lookback)
    out["minervini_c3"] = sma200 > sma200_lagged
    # C4: sma_50 > sma_150 > sma_200
    out["minervini_c4"] = (sma50 > sma150) & (sma150 > sma200)
    # C5: close > sma_50
    out["minervini_c5"] = close > sma50
    # C6: close >= w52_low * C6_W52LOW_MULT (책 임계는 SSOT docstring 참조)
    out["minervini_c6"] = close >= w52l * C6_W52LOW_MULT
    # C7: close >= w52_high * 0.75
    out["minervini_c7"] = close >= w52h * C7_W52HIGH_MULT

    # NaN 보존: 입력 중 하나라도 NaN 이면 조건도 NaN (pandas 비교 결과는 False 이지만, 우리는 NaN 으로)
    for c, cols in [
        ("minervini_c1", [close, sma150, sma200]),
        ("minervini_c2", [sma150, sma200]),
        ("minervini_c3", [sma200, sma200_lagged]),
        ("minervini_c4", [sma50, sma150, sma200]),
        ("minervini_c5", [close, sma50]),
        ("minervini_c6", [close, w52l]),
        ("minervini_c7", [close, w52h]),
    ]:
        # 어느 입력이라도 NaN 인 행은 조건도 NaN (object dtype 으로 변환)
        nan_mask = pd.concat([col.isna() for col in cols], axis=1).any(axis=1)
        out[c] = out[c].astype(object)
        out.loc[nan_mask, c] = np.nan

    # 데이터 결함 가드: w52_low<=0 (저가 0행으로 rolling min 이 0) 이면 C6 평가 불가.
    # close>=0 이 항상 True 라 가짜 통과되는 것을 막고, NaN(cannot-evaluate)으로 둬
    # minervini_pass(c6 IS TRUE AND ...)에서 FALSE → 후보 제외. 데이터 복구 시 자동 복귀.
    # (w52l<=0 비교는 NaN 행을 False 로 둬 위 NaN 보존과 충돌하지 않음.)
    out.loc[w52l <= 0, "minervini_c6"] = np.nan

    return out

File: web/src/data/llm-pipeline-audit/minervini.ts
// Minervini Trend Template 8조건 (spec audit §4)

export interface MinerviniCondition {
  num: number;
  korean: string;
  threshold: string;
  codeRef: string;
  englishOriginal: string;
  note?: string;
}

export const MINERVINI_PASS_FORMULA = `
minervini_pass = (
    minervini_c1 IS TRUE AND minervini_c2 IS TRUE AND
    minervini_c3 IS TRUE AND minervini_c4 IS TRUE AND
    minervini_c5 IS TRUE AND minervini_c6 IS TRUE AND
    minervini_c7 IS TRUE AND (rs_rating >= 70)
)
`.trim();

export const MINERVINI_PASS_REF = "kr_pipeline/indicators/store.py:91-96 (SQL UPDATE SET)";

export const MINERVINI_CONDITIONS: MinerviniCondition[] = [
  {
    num: 1,
    korean: "close > sma_150 AND sma_150 > sma_200",
    threshold: "—",
    codeRef: "minervini.py:27",
    englishOriginal: "Price > MA150 AND MA150 > MA200",
  },
  {
    num: 2,
    korean: "sma_150 > sma_200",
    threshold: "—",
    codeRef: "minervini.py:29",
    englishOriginal: "MA150 > MA200",
  },
  {
    num: 3,
    korean: "오늘 sma_200 > 22거래일 전 sma_200",
    threshold: "22 거래일 (default)",
    codeRef: "minervini.py:31-32",
    englishOriginal: "MA200 trending up for ≥1 month (≥22 trading days)",
    note: "sma_200_lookback=22 default 인자. '연속 상승' 이 아니라 한 번 비교.",
  },
  {
    num: 4,
    korean: "sma_50 > sma_150 AND sma_150 > sma_200",
    threshold: "—",
    codeRef: "minervini.py:34",
    englishOriginal: "MA50 > MA150 > MA200",
  },
  {
    num: 5,
    korean: "close > sma_50",
    threshold: "—",
    codeRef: "minervini.py:36",
    englishOriginal: "Price > MA50",
  },
  {
    num: 6,
    korean: "close ≥ w52_low × 1.25",
    threshold: "1.25×",
    codeRef: "minervini.py:38",
    englishOriginal:
      "Price ≥ 52w-low × 1.25 (TTLC Ch.6 — 최신작) / × 1.30 (TLSMW Ch.5)",
    note: "두 저작 간 버전 차이 — TTLC Ch.6 (+25%) 와 TLSMW Ch.5 (+30%) 모두 책 근거. 우리는 최신작 채택.",
  },
  {
    num: 7,
    korean: "close ≥ w52_high × 0.75",
    threshold: "0.75×",
    codeRef: "minervini.py:40",
    englishOriginal: "Price ≥ 52w-high × 0.75 (within 25% of 52w high)",
  },
  {
    num: 8,
    korean: "rs_rating ≥ 70",
    threshold: "70",
    codeRef: "store.py:91 (SQL UPDATE SET)",
    englishOriginal: "RS Rating ≥ 70",
    note: "RS Rating 개념은 O'Neil HMMS, 임계 70은 Minervini TLSMW Ch.5 — c1-c7 과 함께 minervini_pass 의 8 번째 조건.",
  },
];

export const NAN_POLICY = `
입력 중 하나라도 NaN 이면 조건도 NaN (boolean 강제 안 함, minervini.py:42-55).
SMA 데이터 부족 종목은 minervini_pass = NULL → 게이트 통과 안 함.
`.trim();

export const WEEKLY_MINERVINI_NOTE = `
weekly_indicators 에도 동일 8조건 + minervini_pass 계산 (store.py:168-182).
LLM payload 의 minervini.json 은 일봉 기준 (minervini_detail_builder.py).
`.trim();
